File: tests_opossum/Opossum.IntegrationTests/ReadFromPositionIntegrationTests.cs
using Opossum.Core;
using Opossum.DependencyInjection;
using Opossum.Extensions;

namespace Opossum.IntegrationTests;

/// <summary>
/// Integration tests for the <c>fromPosition</c> parameter of
/// <see cref="IEventStore.ReadAsync"/>.
///
/// Each test creates its own isolated temp directory so tests can run in parallel.
/// </summary>
public sealed class ReadFromPositionIntegrationTests : IDisposable
{
    private readonly string _tempPath;
    private readonly ServiceProvider _serviceProvider;
    private readonly IEventStore _eventStore;

    private record OrderEvent(string OrderId) : IEvent;
    private record ShipmentEvent(string ShipmentId) : IEvent;

    public ReadFromPositionIntegrationTests()
    {
        _tempPath = Path.Combine(
            Path.GetTempPath(),
            "OpossumReadFromPositionTests",
            Guid.NewGuid().ToString());

        var services = new ServiceCollection();
        services.AddOpossum(opt =>
        {
            opt.RootPath = _tempPath;
            opt.FlushEventsImmediately = false;
            opt.UseStore("TestContext");
        });
        _serviceProvider = services.BuildServiceProvider();
        _eventStore = _serviceProvider.GetRequiredService<IEventStore>();
    }

    public void Dispose()
    {
        _serviceProvider.Dispose();
        if (Directory.Exists(_tempPath))
        {
            try
            { Directory.Delete(_tempPath, recursive: true); }
            catch { /* ignore cleanup errors */ }
        }
    }

    // ── helpers ──────────────────────────────────────────────────────────────

    private Task AppendOrderAsync(string orderId) =>
        _eventStore.AppendEventAsync(new OrderEvent(orderId),
            tags: [new Tag("orderId", orderId)]);

    private Task AppendShipmentAsync(string shipmentId) =>
        _eventStore.AppendEventAsync(new ShipmentEvent(shipmentId),
            tags: [new Tag("shipmentId", shipmentId)]);

    // ── tests ─────────────────────────────────────────────────────────────────

    [Fact]
    public async Task ReadAsync_WithoutFromPosition_ReturnsAllEventsAsync()
    {
        await AppendOrderAsync("o1");
        await AppendOrderAsync("o2");
        await AppendOrderAsync("o3");

        var events = await _eventStore.ReadAsync(Query.All(), null);

        Assert.Equal(3, events.Length);
    }

    [Fact]
    public async Task ReadAsync_WithFromPositionZero_ReturnsAllEventsAsync()
    {
        await AppendOrderAsync("o1");
        await AppendOrderAsync("o2");

        var events = await _eventStore.ReadAsync(Query.All(), null, fromPosition: 0);

        Assert.Equal(2, events.Length);
    }

    [Fact]
    public async Task ReadAsync_WithFromPosition_ReturnsOnlyEventsAfterThatPositionAsync()
    {
        // Append 5 events — positions 1 through 5
        for (int i = 1; i <= 5; i++)
            await AppendOrderAsync($"o{i}");

        // Request only events after position 3
        var events = await _eventStore.ReadAsync(Query.All(), null, fromPosition: 3);

        Assert.Equal(2, events.Length);
        Assert.All(events, e => Assert.True(e.Position > 3));
        Assert.Equal(4, events[0].Position);
        Assert.Equal(5, events[1].Position);
    }

    [Fact]
    public async Task ReadAsync_WithFromPositionAtLastEvent_ReturnsEmptyAsync()
    {
        await AppendOrderAsync("o1");
        await AppendOrderAsync("o2");
        await AppendOrderAsync("o3");

        var lastPosition = (await _eventStore.ReadAsync(Query.All(), null)).Max(e => e.Position);

        var events = await _eventStore.ReadAsync(Query.All(), null, fromPosition: lastPosition);

        Assert.Empty(events);
    }

    [Fact]
    public async Task ReadAsync_WithFromPositionAndEventTypeFilter_ReturnsOnlyMatchingTypesAfterPositionAsync()
    {
        await AppendOrderAsync("o1");        // pos 1 – OrderEvent
        await AppendShipmentAsync("s1");    // pos 2 – ShipmentEvent
        await AppendOrderAsync("o2");        // pos 3 – OrderEvent
        await AppendShipmentAsync("s2");    // pos 4 – ShipmentEvent
        await AppendOrderAsync("o3");        // pos 5 – OrderEvent

        var events = await _eventStore.ReadAsync(
            Query.FromEventTypes(nameof(OrderEvent)), null, fromPosition: 2);

        // Only OrderEvent events at positions > 2 → positions 3 and 5
        Assert.Equal(2, events.Length);
        Assert.Equal(3, events[0].Position);
        Assert.Equal(5, events[1].Position);
        Assert.All(events, e => Assert.Equal(nameof(OrderEvent), e.Event.EventType));
    }

    [Fact]
    public async Task ReadAsync_WithFromPositionAndTagFilter_ReturnsOnlyMatchingTagsAfterPositionAsync()
    {
        var tag = new Tag("orderId", "o-special");

        await _eventStore.AppendEventAsync(new OrderEvent("o-special"),
            tags: [tag]);  // pos 1
        await AppendOrderAsync("o-other");  // pos 2
        await _eventStore.AppendEventAsync(new OrderEvent("o-special"),
            tags: [tag]);  // pos 3
        await _eventStore.AppendEventAsync(new OrderEvent("o-special"),
            tags: [tag]);  // pos 4

        var events = await _eventStore.ReadAsync(Query.FromTags(tag), null, fromPosition: 1);

        // Positions 3 and 4 match
        Assert.Equal(2, events.Length);
        Assert.Equal(3, events[0].Position);
        Assert.Equal(4, events[1].Position);
    }

    [Fact]
    public async Task ReadAsync_CanPollIncrementally_SimulatingProjectionDaemonAsync()
    {
        // First batch — simulate existing events
        await AppendOrderAsync("o1"); // pos 1
        await AppendOrderAsync("o2"); // pos 2

        var firstBatch = await _eventStore.ReadAsync(Query.All(), null, fromPosition: 0);
        Assert.Equal(2, firstBatch.Length);

        var checkpoint = firstBatch.Max(e => e.Position); // 2

        // New events arrive
        await AppendOrderAsync("o3"); // pos 3
        await AppendOrderAsync("o4"); // pos 4

        // Poll for only new events using the saved checkpoint
        var secondBatch = await _eventStore.ReadAsync(Query.All(), null, fromPosition: checkpoint);

        Assert.Equal(2, secondBatch.Length);
        Assert.All(secondBatch, e => Assert.True(e.Position > checkpoint));
        Assert.Equal(3, secondBatch[0].Position);
        Assert.Equal(4, secondBatch[1].Position);
    }

    [Fact]
    public async Task ReadAsync_WithFromPositionExtensionMethod_ReturnsOnlyEventsAfterPositionAsync()
    {
        await AppendOrderAsync("o1"); // pos 1
        await AppendOrderAsync("o2"); // pos 2
        await AppendOrderAsync("o3"); // pos 3

        // Use the convenience extension method
        var events = await _eventStore.ReadAsync(Query.All(), fromPosition: 1);

        Assert.Equal(2, events.Length);
        Assert.Equal(2, events[0].Position);
        Assert.Equal(3, events[1].Position);
    }
}

File: tests/read_from_position_integration_tests.rs
mod common;

use common::{FakeClock, InMemoryStorage};
use marmosa::domain::{DomainEvent, EventData, Query, QueryItem, Tag};
use marmosa::event_store::{EventStore, OpossumStore};
use marmosa::extensions::EventStoreExt;
use std::sync::Arc;

fn create_store() -> (
    Arc<InMemoryStorage>,
    OpossumStore<Arc<InMemoryStorage>, FakeClock>,
) {
    let storage = Arc::new(InMemoryStorage::new());
    let clock = FakeClock::new(1000);
    let store = OpossumStore::new(Arc::clone(&storage), clock);
    (storage, store)
}

async fn append_order_async(store: &impl EventStoreExt, order_id: &str) {
    let event = EventData {
        event_id: format!("order-{}", order_id),
        event: DomainEvent {
            event_type: "OrderEvent".to_string(),
            data: format!(r#"{{"OrderId": "{}"}}"#, order_id),
            tags: vec![Tag {
                key: "orderId".to_string(),
                value: order_id.to_string(),
            }],
        },
        metadata: None,
    };
    store.append_single_async(event, None).await.unwrap();
}

async fn append_shipment_async(store: &impl EventStoreExt, shipment_id: &str) {
    let event = EventData {
        event_id: format!("shipment-{}", shipment_id),
        event: DomainEvent {
            event_type: "ShipmentEvent".to_string(),
            data: format!(r#"{{"ShipmentId": "{}"}}"#, shipment_id),
            tags: vec![Tag {
                key: "shipmentId".to_string(),
                value: shipment_id.to_string(),
            }],
        },
        metadata: None,
    };
    store.append_single_async(event, None).await.unwrap();
}

#[tokio::test]
async fn read_async_without_from_position_returns_all_events() {
    let (_, store) = create_store();
    append_order_async(&store, "o1").await; // pos 0
    append_order_async(&store, "o2").await; // pos 1
    append_order_async(&store, "o3").await; // pos 2

    let events = store
        .read_async(Query::all(), None, None, None)
        .await
        .unwrap();
    assert_eq!(events.len(), 3);
}

#[tokio::test]
async fn read_async_with_from_position_returns_only_events_after_that_position() {
    let (_, store) = create_store();
    for i in 1..=5 {
        append_order_async(&store, &format!("o{}", i)).await; // 0, 1, 2, 3, 4
    }

    // Request only events after position 2 (should return 3 and 4)
    let events = store
        .read_async(Query::all(), Some(2), None, None)
        .await
        .unwrap();

    assert_eq!(events.len(), 2);
    assert_eq!(events[0].position, 3);
    assert_eq!(events[1].position, 4); // C# has positions 4 and 5 if 1..=5 created 5 items. 
    // Wait: C# adds positions 0,1,2,3,4.
    // After 2 -> positions 3 and 4. Correct.
}

#[tokio::test]
async fn read_async_with_from_position_at_last_event_returns_empty() {
    let (_, store) = create_store();
    append_order_async(&store, "o1").await;
    append_order_async(&store, "o2").await;
    append_order_async(&store, "o3").await;

    let all_events = store
        .read_async(Query::all(), None, None, None)
        .await
        .unwrap();
    let last_position = all_events.iter().map(|e| e.position).max().unwrap();

    let events = store
        .read_async(Query::all(), Some(last_position), None, None)
        .await
        .unwrap();
    assert!(events.is_empty());
}

#[tokio::test]
async fn read_async_with_from_position_and_event_type_filter_returns_only_matching_types_after_position()
 {
    let (_, store) = create_store();
    append_order_async(&store, "o1").await; // pos 0
    append_shipment_async(&store, "s1").await; // pos 1
    append_order_async(&store, "o2").await; // pos 2
    append_shipment_async(&store, "s2").await; // pos 3
    append_order_async(&store, "o3").await; // pos 4

    let query = Query {
        items: vec![QueryItem {
            event_types: vec!["OrderEvent".to_string()],
            tags: vec![],
        }],
    };

    // After pos 1, we only want OrderEvents -> should be at pos 2 and 4
    let events = store.read_async(query, Some(1), None, None).await.unwrap();

    assert_eq!(events.len(), 2);
    assert_eq!(events[0].position, 2);
    assert_eq!(events[1].position, 4);
    assert!(events.iter().all(|e| e.event.event_type == "OrderEvent"));
}

#[tokio::test]
async fn read_async_with_from_position_and_tag_filter_returns_only_matching_tags_after_position() {
    let (_, store) = create_store();
    let tag = Tag {
        key: "orderId".to_string(),
        value: "o-special".to_string(),
    };

    append_order_async(&store, "o-special").await; // pos 0
    append_order_async(&store, "o-other").await; // pos 1
    append_order_async(&store, "o-special").await; // pos 2
    append_order_async(&store, "o-special").await; // pos 3

    let query = Query {
        items: vec![QueryItem {
            event_types: vec![],
            tags: vec![tag],
        }],
    };

    // After pos 0 -> should return pos 2 and 3
    let events = store.read_async(query, Some(0), None, None).await.unwrap();

    assert_eq!(events.len(), 2);
    assert_eq!(events[0].position, 2);
    assert_eq!(events[1].position, 3);
}

#[tokio::test]
async fn read_async_can_poll_incrementally_simulating_projection_daemon() {
    let (_, store) = create_store();

    // First batch
    append_order_async(&store, "o1").await; // pos 0
    append_order_async(&store, "o2").await; // pos 1

    let first_batch = store
        .read_async(Query::all(), None, None, None)
        .await
        .unwrap();
    assert_eq!(first_batch.len(), 2);

    let checkpoint = first_batch.iter().map(|e| e.position).max().unwrap_or(0); // 1

    // New events
    append_order_async(&store, "o3").await; // pos 2
    append_order_async(&store, "o4").await; // pos 3

    let second_batch = store
        .read_async(Query::all(), Some(checkpoint), None, None)
        .await
        .unwrap();
    assert_eq!(second_batch.len(), 2);
    assert_eq!(second_batch[0].position, 2);
    assert_eq!(second_batch[1].position, 3);
}

#[tokio::test]
async fn read_async_with_from_position_extension_method_returns_only_events_after_position() {
    let (_, store) = create_store();
    append_order_async(&store, "o1").await; // pos 0
    append_order_async(&store, "o2").await; // pos 1
    append_order_async(&store, "o3").await; // pos 2

    let events = store
        .read_async(Query::all(), Some(0), None, None)
        .await
        .unwrap();

    assert_eq!(events.len(), 2);
    assert_eq!(events[0].position, 1);
    assert_eq!(events[1].position, 2);
}
